lib/vega_lite/data.ex

```diff
@@ -327,6 +327,328 @@ defmodule VegaLite.Data do
+
+    vl
+    |> Vl.data_from_values(%{_r: [1]})
+    |> Vl.layers(grid_layers ++ radius_layers ++ data_layers)
+  end
+
+  defp deg_to_rad(angle), do: angle * :math.pi() / 180
+
+  defp polar_angle_layers(opts) do
+    angle_marks_input = opts[:angle_marks]
+    theta_offset = opts[:theta_offset]
+
+    {angle_marks, angle_marks2, theta_offset} =
+      case opts[:direction] do
+        :clockwise ->
+          angle_marks = [0 | Enum.sort(angle_marks_input)]
+          angle_marks2 = tl(angle_marks) ++ [360]
+
+          {angle_marks, angle_marks2, theta_offset + 90}
+
+        :counter_clockwise ->
+          angle_marks = [360 | Enum.sort(angle_marks_input, :desc)]
+          angle_marks2 = tl(angle_marks) ++ [0]
+
+          {Enum.map(angle_marks, &(-&1)), Enum.map(angle_marks2, &(-&1)), -theta_offset + 90}
+      end
+
+    has_zero = 0 in angle_marks_input
+
+    [angle_marks, angle_marks2]
+    |> Enum.zip_with(fn [t, t2] ->
+      is_360 = :math.fmod(t, 360) == 0
+
+      label =
+        if (t != 0 and not is_360) or (t == 0 and has_zero) or
+             (is_360 and not has_zero) do
+          Vl.new()
+          |> Vl.mark(:text,
+            text: to_string(abs(t)) <> "º",
+            theta: "#{deg_to_rad(t + theta_offset)}",
+            radius: [expr: "min(width, height) * 0.55"]
+          )
+        else
+          []
+        end
+
+      theta = deg_to_rad(t + theta_offset)
+      theta2 = deg_to_rad(t2 + theta_offset)
+
+      [
+        Vl.new()
+        |> Vl.mark(:arc,
+          theta: "#{theta}",
+          theta2: "#{theta2}",
+          stroke: opts[:grid_stroke_color],
+          stroke_opacity: opts[:grid_stroke_opacity],
+          opacity: opts[:grid_opacity],
+          color: opts[:grid_color]
+        ),
+        label
+      ]
+    end)
+    |> List.flatten()
+  end
+
+  defp polar_radius_layers(opts) do
+    radius_marks = opts[:radius_marks]
+    max_radius = Enum.max(radius_marks)
+
+    radius_marks_vl =
+      Enum.map(radius_marks, fn r ->
+        Vl.mark(Vl.new(), :arc,
+          radius: [expr: "#{r / max_radius} * min(width, height)/2"],
+          radius2: [expr: "#{r / max_radius} * min(width, height)/2 + 1"],
+          theta: "0",
+          theta2: "#{2 * :math.pi()}",
+          stroke_color: opts[:grid_stroke_color],
+          color: opts[:grid_stroke_color],
+          opacity: opts[:grid_stroke_opacity]
+        )
+      end)
+
+    radius_ruler_vl = [
+      Vl.new()
+      |> Vl.data_from_values(%{
+        r: radius_marks,
+        theta: Enum.map(radius_marks, fn _ -> :math.pi() / 4 end)
+      })
+      |> Vl.mark(:text,
+        color: opts[:grid_stroke_color],
+        radius: [expr: "datum.r  * min(width, height) / (2 * #{max_radius})"],
+        theta: :math.pi() / 2,
+        dy: 10,
+        dx: -10
+      )
+      |> Vl.encode_field(:text, "r", type: :quantitative)
+    ]
+
+    radius_marks_vl ++ radius_ruler_vl
+  end
+
+  defp polar_infer_radius_marks(data) do
```

Comment:
We don't do data processing. And we can't assume data is an enum nor fields in the data. We only assume data implements Table.Reader.

Reply:
This is a deal breaker in terms of usability, but again, can be extracted out.

Reply:
The rest of the preprocessing is done through VegaLite calculations, so it's fine

Reply:
Yeah, it is up to you. I am mostly dropping comments in favor of consistency with the remaining APIs in this module, which I would like to strictly adhere to for now.

Reply:
It's not clear to me.

If we:

- remove the max radius inference
- refactor the data keys so that we don't assume the keys in data
- only accept a single data collection instead of a list of those

Is the feature good to go? Or would it still be in the "can be denied" realm?

Reply:
Then I think we can accept it, if you still think it is useful. But I can't promise other blockers won't be found as you refactor. I can't promise to merge something based on code i can't see.